design-proposals/high-level-plane.md

```diff
@@ -0,0 +1,57 @@
+# High-level plane
+
+Author: Fabian Deutsch \<fdeutsch@redhat.com\>
+
+## Introduction
+
+Today KubeVirt is exposing a pretty low-level interface to virtualization capabilities.
+
+But this low-level API might be too low-level for i.e. UIs to integrate. On the higher level its typically required to set policies, the lower layer then provides the mechanics to implement this policy. The suggested high-level plane is the middle layer, translating policies into the mechanics.
+
+The introduction of a high level plane also tries to prevent that too much logic is pushed into the lower level mechanics.
+
+There will be users for a high-level plane, but we assume that the low-level plane is still relevant, that is the reason why a separate API is suggested, instead of changing the existing one.
+
+### Use-case
+
+The primary focus is to provide a convenient high-level API suitable for integration into UIs.
+
+A couple of aspects will be the focus of this initial part:
+
+* VM abstraction (support stopped VMs, OS flavors, disk abstraction, network abstraction, )
+* VM templating
+* …
+
+## API
+
+TBD
+The API is probably just a more high-level VM Spec for now, it would exist in parallel to the low-level representation.
+Example:
+
+kind: VirtualMachine
+spec:
+  state: down
```

Comment:
@fabiand could you clarify if this proposal is about having add-ons or, having this add-on?

I think we need to have a clear concrete use-case (nothing fancy just a clear sentence) on why we want to have this extra api.
If we don't do that, it is not clear to me what it is for and to distinguish what should be in an add-on and what should go into kubevirt/kubevirt.

Some use-cases which I could think off:

 1. I want to store my VMs somewhere so that I can start them later on
 2. I want to see shut-off VMs, which should be running
 3. I want  to see crashed VMs, which should be running
 4. I want to see running VMs
 5. I want to have flavours or templates to make it easier to re-use VM definitions
 6. I want to browse online and offline VMs in different UIs on different scales

Possible distinction between add-on and core feature:

 1. vm store (add-on)
 2. probably a core feature, since the VM should be running, kubevirt should try to correct that mistake which means events and status updates on the core API. Seeing a difference between the wanted and the actual state should always be possible.
 3. like 2
 4. core feature
 5. Could be an add-on or a core feature
 6. Here both APIs might not be enough, depending on the UI (mobile, full blown app, ...) we probably need different APIs reflecting different usage concepts.